rust/private/rust_analyzer.bzl

```diff
@@ -0,0 +1,205 @@
+        crate_root = "../" * rule_depth + info.crate.root.dirname + "/../"
+
+    if info.build_info != None:
+        crate["env"].update({"OUT_DIR": ctx.attr.exec_root + "/" + info.build_info.out_dir.path})
+        crate["source"] = {
+            # We have to tell rust-analyzer about our out_dir since it's not under the crate root.
+            "include_dirs": [crate_root, ctx.attr.exec_root + "/" + info.build_info.out_dir.path],
+            "exclude_dirs": [],
+        }
+    crate["env"].update(info.env)
+
+    deps = []
+    for dep in info.deps:
+        deps.append({
+            "name": dep.crate.name,
+            "crate": crate_mapping["ID-" + dep.crate.root.path],
+        })
+
+    crate["deps"] = deps
+    crate["cfg"] = info.cfgs
+    crate["target"] = find_toolchain(ctx).target_triple
+    if info.proc_macro_dylib_path != None:
+        crate["proc_macro_dylib_path"] = ctx.attr.exec_root + "/" + info.proc_macro_dylib_path
+    return "ID-" + info.crate.root.path, crate
+
+# This implementation is incomplete because in order to get rustc env vars we
+# would need to actually execute the build graph and gather the output of
+# cargo_build_script rules. This would require a genrule to actually construct
+# the JSON, rather than being able to build it completly in starlark.
+# TODO(djmarcin): Run the cargo_build_scripts to gather env vars correctly.
+def _rust_project_impl(ctx):
+    rust_toolchain = find_toolchain(ctx)
+    crate_mapping = dict()
+
+    output = dict()
+    output["sysroot_src"] = ctx.attr.exec_root + "/" + rust_toolchain.rust_lib.label.workspace_root + "/lib/rustlib/src/library"
+    output["crates"] = []
+
+    # Gather all crates and their dependencies into an array.
+    # Dependencies are referenced by index, so leaves should come first.
+    idx = 0
+    for target in ctx.attr.targets:
+        if RustAnalyzerInfo not in target:
+            continue
+
+        # Add this crate's transitive deps to the crate mapping and output.
+        for dep_info in target[RustAnalyzerInfo].transitive_deps.to_list():
+            crate_id, crate = create_crate(ctx, dep_info, crate_mapping)
+            if crate_id not in crate_mapping:
+                crate_mapping[crate_id] = idx
+                idx += 1
+                output["crates"].append(crate)
+
+        # Add this crate to the crate mapping and output.
+        crate_id, crate = create_crate(ctx, target[RustAnalyzerInfo], crate_mapping)
+        if crate_id not in crate_mapping:
+            crate_mapping[crate_id] = idx
+            idx += 1
+            output["crates"].append(crate)
+
+    ctx.actions.write(output = ctx.outputs.filename, content = struct(**output).to_json())
+
+rust_analyzer = rule(
+    attrs = {
+        "targets": attr.label_list(
+            aspects = [rust_analyzer_aspect],
+            doc = "List of all targets to be included in the index",
+        ),
+        "exec_root": attr.string(
```

Comment:
Very interested in avoiding this. Is there no way to get this infer this information?

Reply:
> The rule requires the user to manually enter their bazel info execution_root in order to locate remote crates and out_dirs. Is there a better way to retrieve this from inside starlark without resorting to genrules? I'd like to leave a genrule rewrite to a later date to unblock the relatively large win this PR brings. 

Can you explain more on why needing `execution_root` is necessary?

Reply:
I agree, this is the biggest issue with this PR. I believe the one real viable alternative is setting a custom `--workspace_status_command` https://docs.bazel.build/versions/master/user-manual.html#flag--workspace_status_command to provide this information via an environment variable, but very interested to know if I missed another way.

Reply:
Maybe @damienmg or @mfarrugi might know?

The thing I want from this PR is to be able to push a change to my repo and have enabled `rust-analyzer` support for anyone with no additional work. Having someone need to explicitly set the output of `bazel info execution_root` sounds like it wouldn't support that. I think `--workspace_status_command` might be something that works? Since I could commit a `.bazelrc` file to the repo to enable these rules. But definitely interested in finding something more contained. I'm trying to find such a solution but am having no luck 😞

Reply:
The reason why it's necessary to get this information is that we must provide a path to various external source files, e.g. the rust sysroot, any third party crates you depend on, etc. All of the paths generated during the action analysis phase are relative to the execution root, and are not available via the bazel-* symlinks, so it's not possible to generate a relative path from the user's source directory to the external sources.

I'm toying with the idea of a simple genrule that can get execution_root and uses sed as a template expander, but it's not portable (e.g. to windows). There is a built-in action `ctx.actions.expand_template`, but unfortunately in our case the substitution would be in an output and that rule requires it to be available as a starlark variable, so I don't think it will work.

Reply:
I should also mention that I'm really not sure how `--workspace_status_command` or even the genfiles solution would work, because you can't recursively run `bazel info` in order to find the workspace at runtime. So, passing the `exec_root` is still the best solution I have that works at the moment, unsatisfying as it is.

Reply:
If there's a path forward with `genrule`, you could use [cmd_ps](https://docs.bazel.build/versions/master/be/general.html#genrule.cmd_ps) or [cmd_bat](https://docs.bazel.build/versions/master/be/general.html#genrule.cmd_bat) to handle the windows specific behavior, right?

Reply:
Ok a few updates here:
1) The `bazel-{workspace}` symlink does point at the execution root, I don't know how I missed that before. I'm not sure this actually helps because I'm not sure I can get the workspace name in a normal (non-repository) rule either, and also because of #2

2) Due to how symlinks work, `bazel-bin/..` is **not** your current directory, it is actually a path in the execution root. For example, if `bazel-bin -> /home/vagrant/.cache/bazel/_bazel_vagrant/71ed060a30795ab0736b520a305b19cd/execroot/__main__/bazel-out/k8-fastbuild/bin` then `bazel-bin/..` is actually `.../execroot/__main__/bazel-out/k8-fastbuild`
This has a couple of consequences. First, the way this PR resolves local source code is wrong, but for whatever reason it doesn't look like this is breaking anything. Second, we could drop the exec_root parameter if we were willing to make the assumption that the execroot is at bazel-bin/../../../, which I _think_ is valid and fairly safe. This is assuming that the generated rust-project.json will be used directly from bazel-bin, which is the current assumption made in this PR.

Reply:
For some reason I don't fully understand yet, switching to entirely relative paths based on this assumption breaks rust-analyzer.

Reply:
This is fixed by adding the gazelle-like utility run target.